rskj-core/src/main/java/co/rsk/core/bc/ParallelizeTransactionHandler.java

```diff
@@ -128,107 +110,118 @@ public short[] getTransactionsPerBucketInOrder() {
-        bucketIdsBySender.add(bucketId);
-        bucketBySender.put(sender, bucketIdsBySender);
+        if (bucket.isSequential()) {
+            bucketBySender.put(sender, bucket);
+        } else {
+            bucketBySender.putIfAbsent(sender, bucket);
+        }
 
-        if (bucketId.equals(this.sequentialBucket.getId())) {
+        if (bucket.equals(getSequentialBucket())) {
             return;
         }
 
         for (ByteArrayWrapper key: newWrittenKeys) {
-            bucketByWrittenKey.putIfAbsent(key, bucketId);
+            bucketByWrittenKey.putIfAbsent(key, bucket);
         }
     }
 
-    private Optional<Set<Short>> getBucketBySender(Transaction tx) {
+    private Optional<TransactionBucket> getBucketBySender(Transaction tx) {
         return Optional.ofNullable(bucketBySender.get(tx.getSender()));
     }
 
-    private Optional<Short> getAvailableBucketWithLessUsedGas(long txGasLimit) {
+    private Optional<TransactionBucket> getAvailableBucketWithLessUsedGas(long txGasLimit) {
         long gasUsed = Long.MAX_VALUE;
-        Optional<Short> bucketId =  Optional.empty();
+        Optional<TransactionBucket> bucketCandidate =  Optional.empty();
 
-        for (TransactionBucket bucket : parallelBuckets) {
-            if (bucket.hasGasAvailable(txGasLimit)) {
-                if (bucket.getGasUsed() < gasUsed) {
-                    bucketId = Optional.of(bucket.getId());
+        for (TransactionBucket bucket : buckets) {
+            if (!bucket.isSequential() && bucket.hasGasAvailable(txGasLimit) && bucket.getGasUsed() < gasUsed) {
+                    bucketCandidate = Optional.of(bucket);
                     gasUsed = bucket.getGasUsed();
-                }
             }
         }
-        return bucketId;
+
+        return bucketCandidate;
     }
 
 
-    private Set<Short> getBucketCandidates(Transaction tx, Set<ByteArrayWrapper> newReadKeys, Set<ByteArrayWrapper> newWrittenKeys) {
-        Set<Short> bucketCandidates = new HashSet<>();
+    private TransactionBucket getBucketCandidates(Transaction tx, Set<ByteArrayWrapper> newReadKeys, Set<ByteArrayWrapper> newWrittenKeys) {
+        Optional<TransactionBucket> bucketCandidate = getBucketBySender(tx);
 
-        getBucketBySender(tx).ifPresent(bucketCandidates::addAll);
+        if (bucketCandidate.isPresent() && bucketCandidate.get().isSequential()) {
+            return getSequentialBucket();
+        }
 
         // read - written
         for (ByteArrayWrapper newReadKey : newReadKeys) {
             if (bucketByWrittenKey.containsKey(newReadKey)) {
-                bucketCandidates.add(bucketByWrittenKey.get(newReadKey));
+                TransactionBucket bucket = bucketByWrittenKey.get(newReadKey);
+
+                if (bucketCandidate.isPresent() && !bucketCandidate.get().equals(bucket)) {
+                   return getSequentialBucket();
+                } else {
```

Comment:
I think this else block can be removed or replaced by `else if (!bucketCandidate.isPresent())`